apps/serdes/Deserializer.cpp

```diff
@@ -0,0 +1,608 @@
+        auto then_case = deserialize_stmt(if_then_else_stmt->then_case_type(), if_then_else_stmt->then_case());
+        auto else_case = deserialize_stmt(if_then_else_stmt->else_case_type(), if_then_else_stmt->else_case());
+        return Halide::Internal::IfThenElse::make(condition, then_case, else_case);
+    }
+    case Halide::Serdes::Stmt_Evaluate: {
+        const Halide::Serdes::Evaluate *evaluate_stmt = (const Halide::Serdes::Evaluate *)stmt;
+        auto value = deserialize_expr(evaluate_stmt->value_type(), evaluate_stmt->value());
+        return Halide::Internal::Evaluate::make(value);
+    }
+    case Halide::Serdes::Stmt_Prefetch: {
+        const Halide::Serdes::Prefetch *prefetch_stmt = (const Halide::Serdes::Prefetch *)stmt;
+        auto name = deserialize_string(prefetch_stmt->name());
+        std::vector<Type> types;
+        for (const auto &type : *prefetch_stmt->types()) {
+            types.push_back(deserialize_type(type));
+        }
+        std::vector<Range> bounds;
+        for (const auto &bound : *prefetch_stmt->bounds()) {
+            bounds.push_back(deserialize_range(bound));
+        }
+        auto condition = deserialize_expr(prefetch_stmt->condition_type(), prefetch_stmt->condition());
+        auto body = deserialize_stmt(prefetch_stmt->body_type(), prefetch_stmt->body());
+        return Halide::Internal::Prefetch::make(name, types, bounds,
+                                                Halide::Internal::PrefetchDirective(),
+                                                condition, body);
+    }
+    case Halide::Serdes::Stmt_Acquire: {
+        const Halide::Serdes::Acquire *acquire_stmt = (const Halide::Serdes::Acquire *)stmt;
+        auto semaphore = deserialize_expr(acquire_stmt->semaphore_type(), acquire_stmt->semaphore());
+        auto count = deserialize_expr(acquire_stmt->count_type(), acquire_stmt->count());
+        auto body = deserialize_stmt(acquire_stmt->body_type(), acquire_stmt->body());
+        return Halide::Internal::Acquire::make(semaphore, count, body);
+    }
+    case Halide::Serdes::Stmt_Fork: {
+        const Halide::Serdes::Fork *fork_stmt = (const Halide::Serdes::Fork *)stmt;
+        auto first = deserialize_stmt(fork_stmt->first_type(), fork_stmt->first());
+        auto rest = deserialize_stmt(fork_stmt->rest_type(), fork_stmt->rest());
+        return Halide::Internal::Fork::make(first, rest);
+    }
+    case Halide::Serdes::Stmt_Atomic: {
+        const Halide::Serdes::Atomic *atomic_stmt = (const Halide::Serdes::Atomic *)stmt;
+        auto producer_name = deserialize_string(atomic_stmt->producer_name());
+        auto mutex_name = deserialize_string(atomic_stmt->mutex_name());
+        auto body = deserialize_stmt(atomic_stmt->body_type(), atomic_stmt->body());
+        return Halide::Internal::Atomic::make(producer_name, mutex_name, body);
+    }
+    default:
+        std::cerr << "unknown type code " << type_code << "\n";
```

Comment:
Don't write to `stderr` directly from within libHalide (some usages bottleneck our error output, e.g. in the Python bindings). 

If this is a fatal error, use `user_assert()`; if it's just recoverable (but needs a warning), use `user_warning`.